### PR TITLE
V3 combine lua-binding, jsb, and simulator

### PR DESCRIPTION
**Pull request by xiaofeng11**
Because simulator is base on libluacocos2d and libjscocos2d, so we combine them in one PR. 

There hasn't any functional modify in this PR, all changes in this PR is make libluacocos2d, libjscocos2d and libsimulator can be compiled as dynamic library and export functions needed by cocos studio. We have add function export declare to related function and modify some function implement method in simulator. In previous version, some function implement in simulator project, but call in libsimulator project, this cause libsimulator can not compiled as a dynamic library.

@ricardoquesada @dualface @minggo @pandamicro Please review this PR.
@linshun Does there still has somebody can help review simulator related modifies? Please help to assign them, Thanks.

### CODE REVIEW

File: cocos/scripting/js-bindings/manual/js-BindingsExport.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+            #define CC_JS_DLL     __declspec(dllexport)
+        #else         /* use a DLL library */
+            #define CC_JS_DLL     __declspec(dllimport)
+        #endif
+    #endif
+
+    /* Define NULL pointer value */
+    #ifndef NULL
+        #ifdef __cplusplus
+            #define NULL    0
```
Comment (minggo) [inline cocos/scripting/js-bindings/manual/js-BindingsExport.h]: NULL -> nullptr

File: cocos/scripting/js-bindings/manual/js_module_register.cpp [hunk 2]
```diff
@@ -1,7 +1,5 @@
-#ifndef __JS_TEMPLATE_RUNTIME_FRAMEWORKS_RUNTIME_SRC_CLASSES_JS_MODULE_REGISTER_H__
-#define __JS_TEMPLATE_RUNTIME_FRAMEWORKS_RUNTIME_SRC_CLASSES_JS_MODULE_REGISTER_H__
+#include "js_module_register.h"
 
-#include "scripting/js-bindings/manual/ScriptingCore.h"
```
Comment (minggo) [inline cocos/scripting/js-bindings/manual/js_module_register.cpp:4]: Why move this file to cocos?
And why change .h to .cpp?

File: tools/simulator/libsimulator/lib/runtime/Widget_mac.h [hunk 3]
```diff
@@ -0,0 +1,12 @@
+//
+//  Widget_mac.h
+//  Simulator
+//
+//
+
+#ifndef libsimulator_studio_Widget_mac_h
+#define libsimulator_studio_Widget_mac_h
```
Comment (minggo) [inline tools/simulator/libsimulator/lib/runtime/Widget_mac.h]: Use `#pragma once` instead.

File: tools/simulator/libsimulator/lib/runtime/Widget_mac.h [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+std::string getCurAppName(void);
```
Comment (minggo) [inline tools/simulator/libsimulator/lib/runtime/Widget_mac.h]: Don't need `void`.

File: cocos/scripting/js-bindings/manual/js_module_register.cpp (hunk 2 above)
Comment (xiaofeng11) [inline cocos/scripting/js-bindings/manual/js_module_register.cpp:4]: @minggo  Not change .h to .cpp, we add the .cpp, and move the implement from .h to .cpp, so simulator can only include the .h file, we can decrease the function need to be export.
As this change , both simulator and cocostudio can include this .h file to build as dylib.

File: cocos/scripting/lua-bindings/manual/lua_module_register.cpp [hunk 5]
```diff
@@ -13,7 +12,8 @@
 #include "physics3d/lua_cocos2dx_physics3d_manual.h"
 #include "navmesh/lua_cocos2dx_navmesh_manual.h"
 
-static int lua_module_register(lua_State* L)
```
Comment (xiaofeng11) [inline cocos/scripting/lua-bindings/manual/lua_module_register.cpp:16]: This is file is the same as js_module_register.h

**Comment by ricardoquesada** [comment]
@xiaofeng11 @linshun 
I think the owner of "bindings" should review this PR, and if the owner needs help we can help him review it. But I don't think four people should review it... it is a waste of time.

I didn't find anything strange in the PR, but I'm not familiar with the runtime code, so I wouldn't know.

**Comment by xiaofeng11** [comment]
[ci rebuild]

File: tools/simulator/frameworks/runtime-src/Classes/ide-support/RuntimeJsImpl.cpp [hunk 6]
```diff
@@ -146,10 +147,10 @@ void register_FileUtils(JSContext *cx, JS::HandleObject global)
...
         ns.set(nsval.toObjectOrNull());
     }
-
-    JS::RootedObject proto(cx, jsb_cocos2d_FileUtils_prototype);
-    JS_DefineFunction(cx, proto, "addSearchPath", runtime_FileUtils_addSearchPath, 1, JSPROP_PERMANENT | JSPROP_ENUMERATE);
-    JS_DefineFunction(cx, proto, "setSearchPaths", runtime_FileUtils_setSearchPaths, 1, JSPROP_PERMANENT | JSPROP_ENUMERATE);
+    JSObject* fileUtils = get_jsb_cocos2d_FileUtils_prototype();
+   JS::RootedObject proto(cx, fileUtils);
+   JS_DefineFunction(cx, proto, "addSearchPath", runtime_FileUtils_addSearchPath, 1, JSPROP_PERMANENT | JSPROP_ENUMERATE);
+   JS_DefineFunction(cx, proto, "setSearchPaths", runtime_FileUtils_setSearchPaths, 1, JSPROP_PERMANENT | JSPROP_ENUMERATE);
```
Comment (pandamicro) [inline tools/simulator/frameworks/runtime-src/Classes/ide-support/RuntimeJsImpl.cpp]: Indent wrong

File: cocos/scripting/js-bindings/manual/js_module_register.cpp [hunk 7]
```diff
@@ -121,5 +119,7 @@ int js_module_register()
     return 1;
 }
 
-#endif  // __JS_TEMPLATE_RUNTIME_FRAMEWORKS_RUNTIME_SRC_CLASSES_JS_MODULE_REGISTER_H__
-
+JSObject* get_jsb_cocos2d_FileUtils_prototype()
```
Comment (pandamicro) [inline cocos/scripting/js-bindings/manual/js_module_register.cpp]:
```
JS::HandleObject get_jsb_cocos2d_FileUtils_prototype()
{
    JS::RootedObject fileUtilsProto(ScriptingCore::getInstance()->getGlobalContext(), jsb_cocos2d_FileUtils_prototype);
    return fileUtilsProto;
}
```

File: tools/simulator/frameworks/runtime-src/Classes/ide-support/RuntimeJsImpl.cpp [hunk 8]
```diff
@@ -146,10 +147,10 @@ void register_FileUtils(JSContext *cx, JS::HandleObject global)
     } else {
         ns.set(nsval.toObjectOrNull());
     }
-
-    JS::RootedObject proto(cx, jsb_cocos2d_FileUtils_prototype);
-    JS_DefineFunction(cx, proto, "addSearchPath", runtime_FileUtils_addSearchPath, 1, JSPROP_PERMANENT | JSPROP_ENUMERATE);
-    JS_DefineFunction(cx, proto, "setSearchPaths", runtime_FileUtils_setSearchPaths, 1, JSPROP_PERMANENT | JSPROP_ENUMERATE);
+    JSObject* fileUtils = get_jsb_cocos2d_FileUtils_prototype();
+   JS::RootedObject proto(cx, fileUtils);
```
Comment (pandamicro) [inline tools/simulator/frameworks/runtime-src/Classes/ide-support/RuntimeJsImpl.cpp]: JS::RootedObject proto(cx, JS::RootedObject proto(cx, fileUtils));

Don't use JSObject *

**Comment by xiaofeng11** [comment]
[ci rebuild]

**Comment by xiaofeng11** [comment]
squash commint and move to https://github.com/cocos2d/cocos2d-x/pull/14534

**Pull request closed without merging**